File: mmupnp/src/main/java/net/mm2d/upnp/internal/thread/TaskExecutors.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.thread;

import net.mm2d.upnp.TaskExecutor;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public class TaskExecutors {
    @Nonnull
    private final TaskExecutor mCallbackExecutor;
    @Nonnull
    private final TaskExecutor mIoExecutor;
    @Nonnull
    private final TaskExecutor mManagerExecutor;
    @Nonnull
    private final TaskExecutor mServerExecutor;

    public TaskExecutors() {
        this(null);
    }

    public TaskExecutors(
            @Nullable final TaskExecutor callback) {
        this(callback, null);
    }

    // VisibleForTesting
    TaskExecutors(
            @Nullable final TaskExecutor callback,
            @Nullable final TaskExecutor io) {
        mCallbackExecutor = callback != null ? callback : ExecutorFactory.createCallback();
        mIoExecutor = io != null ? io : ExecutorFactory.createIo();
        mManagerExecutor = ExecutorFactory.createManager();
        mServerExecutor = ExecutorFactory.createServer();
    }

    public boolean callback(@Nonnull final Runnable task) {
        return mCallbackExecutor.execute(task);
    }

    public boolean io(@Nonnull final Runnable task) {
        return mIoExecutor.execute(task);
    }

    public boolean manager(@Nonnull final Runnable task) {
        return mManagerExecutor.execute(task);
    }

    public boolean server(@Nonnull final Runnable task) {
        return mServerExecutor.execute(task);
    }

    public void terminate() {
        mCallbackExecutor.terminate();
        mIoExecutor.terminate();
        mManagerExecutor.terminate();
        mServerExecutor.terminate();
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/thread/TaskExecutors.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.thread

import net.mm2d.upnp.TaskExecutor

internal class TaskExecutors(
        callback: TaskExecutor? = null,
        io: TaskExecutor? = null
) {
    private val callbackExecutor = callback ?: ExecutorFactory.createCallback()
    private val ioExecutor = io ?: ExecutorFactory.createIo()
    private val managerExecutor = ExecutorFactory.createManager()
    private val serverExecutor = ExecutorFactory.createServer()

    fun callback(task: Runnable): Boolean {
        return callbackExecutor.execute(task)
    }

    fun callback(task: () -> Unit): Boolean = callback(Runnable { task() })

    fun io(task: Runnable): Boolean {
        return ioExecutor.execute(task)
    }

    fun io(task: () -> Unit): Boolean = io(Runnable { task() })

    fun manager(task: Runnable): Boolean {
        return managerExecutor.execute(task)
    }

    fun server(task: Runnable): Boolean {
        return serverExecutor.execute(task)
    }

    fun terminate() {
        callbackExecutor.terminate()
        ioExecutor.terminate()
        managerExecutor.terminate()
        serverExecutor.terminate()
    }
}
